serverless/resource_deployment.go

```diff
@@ -0,0 +1,273 @@
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+type serverlessConfig struct {
+	Service string
+}
+
+func getServerlessConfig(configDir string, serverlessBinDir string) ([]byte, error) {
+	cmd := exec.Command(getServerlessBin(serverlessBinDir), "print", "--format", "json")
+	cmd.Dir = configDir
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return []byte{}, fmt.Errorf("%v\n%w", string(output), err)
+	}
+
+	return output, nil
+}
+
+func getServiceName(configJson []byte) (string, error) {
+	config := serverlessConfig{}
+	err := json.Unmarshal([]byte(configJson), &config)
+
+	return config.Service, err
+}
+
+func hashServerlessDir(configDir string, packagePath string, configJson []byte) (string, error) {
+	absolutePackagePath := filepath.Join(configDir, packagePath)
+	zipPath := filepath.Join(absolutePackagePath, "sls-provider.zip")
+
+	zipHash, err := dirhash.HashZip(zipPath, dirhash.Hash1)
+	if err != nil {
+		return "", err
+	}
+
+	configHashBytes := sha256.Sum256(configJson)
+	configHash := hex.EncodeToString(configHashBytes[:])
+
+	return strings.Join([]string{zipHash, configHash}, "-"), nil
+}
+
+func getServerlessBin(binPath string) string {
+	suffix := ""
+	if runtime.GOOS == "windows" {
+		suffix = ".cmd"
+	}
+	return filepath.Join(binPath, fmt.Sprintf("serverless%s", suffix))
+}
+
+type serverlessParams struct {
+	command          string
+	serverlessBinDir string
+	configDir        string
+	packageDir       string
+	stage            string
+	args             []interface{}
+}
+
+func runServerless(params *serverlessParams) error {
+	stringArgs := make([]string, len(params.args))
+	for i, v := range stringArgs {
+		stringArgs[i] = fmt.Sprint(v)
+	}
+
+	requiredArgs := []string{
+		params.command,
+		"-s",
+		params.stage,
+	}
+
+	if params.command == "deploy" || params.command == "package" {
+		requiredArgs = append(requiredArgs, "-p", params.packageDir)
+	}
+
+	stringArgs = append(
+		requiredArgs,
+		stringArgs...,
+	)
+
+	cmd := exec.Command(getServerlessBin(params.serverlessBinDir), stringArgs...)
+	cmd.Dir = params.configDir
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("%v\n%w", string(output), err)
+	}
+
+	return nil
+}
+
+func resourceDeployment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDeploymentCreate,
+		Read:   resourceDeploymentRead,
+		Update: resourceDeploymentUpdate,
+		Delete: resourceDeploymentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"config_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"serverless_bin_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"package_dir": &schema.Schema{
```

Comment:
What is `package_dir`? Can we get a comment here?

Reply:
Added comment!